lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,234 @@
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
+        )
+        _blocks_json = json.dumps(
+            [
+                block.to_dict()
+                for block in sorted_blocks
+            ]
+        )
+        hashed = md5(_blocks_json).hexdigest()
+        model, _ = cls.objects.get_or_create(hashed=hashed, defaults={'_blocks_json': _blocks_json})
+        return model
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as an array of BlockRecords. If for some reason block_json
+        is not parsable, the json error will bubble up.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [
+            BlockRecord(data['weight'], data['max_score'], data['block_key'])
+            for data in block_dicts
+        ]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocks model. Create a new instance."
+        )
+
+
+class BlockRecord(object):
+    """
+    An object encapsulating all relevant information for representing a block at the time it was used for grade
+    calculation.
+    """
+
+    def __init__(self, weight, max_score, locator):
+        """
+        Creates a BlockRecord.
+
+        Params:
+            weight (float)
+            max_score (float)
+            locator (BlockUsageLocator or string)
+        """
+        self.weight = weight
+        self.max_score = max_score
+        # pylint: disable=protected_access
+        self.block_key = locator._to_string() if isinstance(locator, BlockUsageLocator) else locator
+
+    def to_dict(self):
+        """
+        Serialize this object to a dict object.
+        """
+        return {
+            'weight': self.weight,
+            'max_score': self.max_score,
+            'block_key': self.block_key,
+        }
+
+    @property
+    def locator(self):
+        """
+        Convenience property allowing a BlockUsageLocator to be returned from this object's serialized data.
+        """
+        return BlockUsageLocator._from_string(self.block_key)  # pylint: disable=protected_access
+
+
+class PersistentSubsectionGrade(TimeStampedModel):
+    """
+    A django model tracking persistent grades at the subsection level.
+
+    TODO: here are the other query patterns listed in the ticket that are not currently used. Should any of these
+    indices be built?
+        user_id, course_id, content_type, is_valid
+        course_id, edit-timestamp
+        edit-timestamp
```

Comment:
Adding indexes will lock the table. Agreed that we can punt on this while the feature is turned off. I'd caution against adding analytics use-cases though. We have some really big indexes on CSM because of that, and many of those queries are best served by export into a separate schema entirely. I think we should focus on the transactional use cases in these models.